crypto/machine.go

```diff
@@ -180,15 +190,192 @@ func (mach *OlmMachine) HandleToDeviceEvent(evt *event.Event) {
+func (mach *OlmMachine) SendEncryptedToDevice(userID id.UserID, deviceID id.DeviceID, content event.Content) error {
+	// get device identity
+	device, err := mach.CryptoStore.GetDevice(userID, deviceID)
+	if err != nil {
+		return err
+	} else if device == nil {
+		// fetch if not found
+		usersToDevices := mach.fetchKeys([]id.UserID{userID}, "", true)
+		if devices, ok := usersToDevices[userID]; ok {
+			if device, ok = devices[deviceID]; !ok {
+				return errors.Errorf("Failed to get identity for device %v", deviceID)
+			}
+		} else {
+			return errors.Errorf("Error fetching devices for user %v", userID)
+		}
+	}
+	// create outbound sessions if missing
+	if err := mach.createOutboundSessions(map[id.UserID]map[id.DeviceID]*DeviceIdentity{
+		userID: {
+			deviceID: device,
+		},
+	}); err != nil {
+		return err
+	}
+
+	// get Olm session
+	olmSess, err := mach.CryptoStore.GetLatestSession(device.IdentityKey)
+	if err != nil {
+		return err
+	}
+	if olmSess == nil {
+		return err
+	}
+
+	encrypted := mach.encryptOlmEvent(olmSess, device, event.ToDeviceForwardedRoomKey, content)
+	encryptedContent := &event.Content{Parsed: &encrypted}
+
+	_, err = mach.Client.SendToDevice(event.ToDeviceEncrypted,
+		&mautrix.ReqSendToDevice{
+			Messages: map[id.UserID]map[id.DeviceID]*event.Content{
+				userID: {
+					deviceID: encryptedContent,
+				},
+			},
+		},
+	)
+
+	return err
+}
+
+// RequestRoomKey sends a key request for a room to the current user's devices. If the context is cancelled, then so is the key request.
+// Returns a bool channel that will get notified either when the key is received or the request is cancelled.
+func (mach *OlmMachine) RequestRoomKey(ctx context.Context, toUser id.UserID, toDevice id.DeviceID,
+	roomID id.RoomID, senderKey id.SenderKey, sessionID id.SessionID) (chan bool, error) {
+
+	requestID := strconv.Itoa(rand.Int())
+	keyResponseReceived := make(chan struct{})
+	// store the channel where we will be notified about responses for this session ID
+	mach.roomKeyRequestFilled.Store(sessionID, keyResponseReceived)
+	// request the keys for this session
+	reqEvtContent := &event.Content{
+		Parsed: event.RoomKeyRequestEventContent{
+			Action: event.KeyRequestActionRequest,
+			Body: event.RequestedKeyInfo{
+				Algorithm: id.AlgorithmMegolmV1,
+				RoomID:    roomID,
+				SenderKey: senderKey,
+				SessionID: sessionID,
+			},
+			RequestID:          requestID,
+			RequestingDeviceID: mach.Client.DeviceID,
+		},
+	}
+
+	toDeviceReq := &mautrix.ReqSendToDevice{
+		Messages: map[id.UserID]map[id.DeviceID]*event.Content{
+			toUser: {
+				toDevice: reqEvtContent,
+			},
+		},
+	}
+	// send messages to the devices
+	if _, err := mach.Client.SendToDevice(event.ToDeviceRoomKeyRequest, toDeviceReq); err != nil {
+		return nil, err
+	}
+	resChan := make(chan bool, 1)
+	go func() {
+		// when exiting, send a message to all devices cancelling this key request
+		defer func() {
```

Comment:
Is there any reason for this to be a `defer` rather than just at the end of the function? It doesn't look like anything should panic and there aren't any early returns.

Reply:
A defer made more sense in my original design of this function tbh (when I thought that we might receive cancellation messages from many devices, before reading that if they don't want to share they just ignore the message). I guess I just like defers.

Moved after the `select`.